File: leetcode/medium/Q142_LinkedListCycleII.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        node = head
        visit = []
        while node:
            if node in visit:
                return node
            visit.append(node)
            node = node.next

File: src/medium/Q142_LinkedListCycleII.java
package medium;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class Q142_LinkedListCycleII {
     class ListNode {
         int val;
         ListNode next;
         ListNode(int x) {
             val = x;
             next = null;
         }
     }
    public class Solution {
        public ListNode detectCycle(ListNode head) {
            Set<ListNode> nodeSet = new HashSet<>();
            ListNode node = head;
            while(node != null) {
                if(nodeSet.contains(node)){
                    return node;
                }
                nodeSet.add(node);
                node = node.next;

            }
            return null;
        }
    }
}
